### PR TITLE
Add interfaces for Applicant Service

### PR DESCRIPTION
**Pull request by bion**
### Description
Please explain the changes you made here.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
works toward #100

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/applicant/Block.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package services.applicant;
+
+import com.google.auto.value.AutoValue;
+
+/** Represents a block in the context of a specific user's application. */
+@AutoValue
+public abstract class Block {
+
+  /** The block's ID. Note this is different from the BlockDefinition's ID. */
+  abstract long id();
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/services/applicant/Block.java:10]: What does block id represent?

Comment (natsid) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:10]: It seems like repeated fields will require `Block`s to have their own set of IDs, distinct from those of `BlockDefinition`. 😕 

`id` will help us map URLs to the right block and vice versa. We haven't thought through exactly how/when we'll generate these `id`s yet though.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/** Defines the interface facade for Applicant service */
+interface ApplicantService {
+
+  /**
+   * Performs a set of updates to the applicant's ApplicantData. Updates are atomic i.e. if any of
+   * them fail validation none of them will be written. programId is used to construct the {@code
+   * ReadOnlyApplicantService} provided in the return value.
+   */
+  CompletionStage<ErrorAnd<ReadOnlyApplicantService, UpdateError>> update(
+      long applicantId, long programId, ImmutableSet<Update> updates);
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:17]: issue: I don't see the `Update` class in this PR. Should we add it or is it fine to add it later?

File: universal-application-tool-0.0.1/app/services/ErrorAnd.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    this.result = checkNotNull(result);
+  }
+
+  /** Constructor for the success case. */
+  public ErrorAnd(T result) {
+    this.result = checkNotNull(result);
+    this.errors = ImmutableSet.of();
+  }
+
+  /** True if there is a result */
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/ErrorAnd.java:34]: nitpick (non-blocking): For consistency with the other docstrings, consider rewording to "Returns true if..."

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.common.collect.ImmutableSet;
+import java.util.concurrent.CompletionStage;
+import models.Applicant;
+import services.ErrorAnd;
+
+/** Defines the interface facade for Applicant service */
+interface ApplicantService {
+
+  /**
+   * Performs a set of updates to the applicant's ApplicantData. Updates are atomic i.e. if any of
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:12]: praise: Thanks for calling out the atomicity of `updates`.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+   * them fail validation none of them will be written. programId is used to construct the {@code
+   * ReadOnlyApplicantService} provided in the return value.
+   */
+  CompletionStage<ErrorAnd<ReadOnlyApplicantService, UpdateError>> update(
+      long applicantId, long programId, ImmutableSet<Update> updates);
+
+  /** Creates a new Applicant for a given user. */
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:19]: suggestion (non-blocking): Is it worth calling out "at the latest version" here? That may just be an implementation detail of `ReadOnlyAppliantService` though.

File: universal-application-tool-0.0.1/app/services/applicant/Block.java (hunk 1 above)
Comment (natsid) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:10]: suggestion (non-blocking): Consider expanding a bit on why these `id`s are different in the docstring.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.concurrent.CompletionStage;
+import models.Applicant;
+import services.ErrorAnd;
+
+/** Defines the interface facade for Applicant service */
+interface ApplicantService {
+
+  /**
+   * Performs a set of updates to the applicant's ApplicantData. Updates are atomic i.e. if any of
+   * them fail validation none of them will be written. programId is used to construct the {@code
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:13]: question: Why not use `{@link ReadOnlyApplicantService}` here? Just trying to learn how to use javadoc fanciness.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+   * ReadOnlyApplicantService} provided in the return value.
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:14]: issue (blocking): Update all references to `ReadOnlyApplicantService` ➡️  `ReadOnlyApplicantProgramService`

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 6 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:13]: Out of sheer laziness and not wanting to import it 😂.. It's a new day. I'll change it 👍🏻

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 2 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:17]: Oh I need to add it, thanks!

File: universal-application-tool-0.0.1/app/services/ErrorAnd.java (hunk 3 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/ErrorAnd.java:34]: done

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 6 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:13]: done

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 7 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:14]: done

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 2 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:17]: done

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java (hunk 5 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:19]: Good call. It's important to the semantics of the method so I think it's right to include.

File: universal-application-tool-0.0.1/app/services/applicant/Block.java (hunk 1 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/Block.java:10]: done

**Review by natsid — APPROVED** [review]

**Pull request closed without merging**